src/pheval/post_processing/post_processing.py

```diff
@@ -228,26 +229,57 @@ def sort_pheval_results(self) -> [PhEvalResult]:
         )
 
 
-def _rank_pheval_result(pheval_result: [PhEvalResult], sort_order: SortOrder) -> pd.DataFrame:
-    """
-    Rank PhEval results post-processed from tool-specific output, managing tied scores (ex aequo)
+class ResultRanker:
+    def __init__(self, pheval_result: List[PhEvalResult], sort_order: SortOrder):
+        """
+        Initialise the PhEvalRanker.
+        Args:
+            pheval_result (List[PhEvalResult]): PhEval results to rank.
+            sort_order (SortOrder): Sorting order based on which ranking is performed.
+        """
+        self.pheval_result = pheval_result
+        self.sort_order = sort_order
+        self.ascending = sort_order == SortOrder.ASCENDING
 
-    Args:
-        pheval_result ([PhEvalResult]): PhEval results obtained from tool-specific output
-        sort_order (SortOrder): Sorting order based on which ranking is performed
+    def rank(self) -> pd.DataFrame:
+        """
+        Rank PhEval results, managing tied scores (ex aequo) and handling grouping_id if present.
 
-    Returns:
-        pd.DataFrame : Ranked PhEval results with tied scores managed
+        Returns:
+            pd.DataFrame : Ranked PhEval results with tied scores managed.
+        """
+        pheval_result_df = pd.DataFrame([data.__dict__ for data in self.pheval_result])
 
-    Raises:
-        ValueError: If an incompatible PhEval result type is encountered
-    """
-    pheval_result_df = pd.DataFrame([data.__dict__ for data in pheval_result])
-    if sort_order == SortOrder.ASCENDING:
-        pheval_result_df["rank"] = pheval_result_df["score"].rank(method="max", ascending=True)
-    elif sort_order == SortOrder.DESCENDING:
-        pheval_result_df["rank"] = pheval_result_df["score"].rank(method="max", ascending=False)
-    return pheval_result_df
+        if self._has_valid_grouping_id(pheval_result_df):
```

Comment:
If there are `None`s mixed with grouping_ids in the dataFrame, shouldn't this be an error state? This only checks for the presence of any `None` and then just uses `_rank_without_grouping_id` if there is a `None`. Surely the "grouping_id" column must either be all `None` or all some otherwise the grouping won't work?

Reply:
Happy to go with whatever you think is best practice here. My idea was that if there was `None` mixed I would just default the ranking behaviour ignoring the missing ones as I didn't want the whole process to finish early and continue doing the next ones. But if you think it is best to throw an error then I can change it to that

Reply:
I think it should result in an error as silently ignoring the grouping id isn't a nice thing to do. It needs to be made clear that when a grouping id is used then all results should have one. This isn't made clear enough in the docs.

I'm going to merge this anyway as this will need changing when migrating over to a parquet and polars solution.